### PR TITLE
[FLINK-5980] Expose max-parallelism value in RuntimeContext.

### PR DESCRIPTION
**Pull request by ifndef-SleePy**
Add new method named getMaxParallelismOfSubtasks in RuntimeContext.

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java [hunk 1]
```diff
@@ -85,6 +85,13 @@
 	int getIndexOfThisSubtask();
 
 	/**
+	 * Gets the number of max-parallelism with which the parallel task runs.
+	 *
+	 * @return The max-parallelism with which the parallel task runs.
+	 */
+	int getMaxParallelismOfSubtasks();
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java]: should this maybe be called ```getMaxNumberOfParallelSubtasks``` to be consistent with other methods?

File: flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java [hunk 2]
```diff
@@ -85,6 +85,13 @@
 	int getIndexOfThisSubtask();
 
 	/**
+	 * Gets the number of max-parallelism with which the parallel task runs.
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java]: can we move this method right below ```getNumberOfParallelSubtasks```?

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java [hunk 3]
```diff
@@ -100,7 +100,12 @@ public int getNumberOfParallelSubtasks() {
 	public int getIndexOfThisSubtask() {
 		return taskInfo.getIndexOfThisSubtask();
 	}
-	
+
+	@Override
+	public int getMaxParallelismOfSubtasks() {
+		// Number of KeyGroups is same with max-parallelism currently.
+		return taskInfo.getNumberOfKeyGroups();
+	}
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java]: misssing empty line

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/async/RichAsyncFunction.java [hunk 4]
```diff
@@ -125,6 +125,10 @@ public int getIndexOfThisSubtask() {
 		}
 
 		@Override
+		public int getMaxParallelismOfSubtasks() {
+			return runtimeContext.getMaxParallelismOfSubtasks();
+		}
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/async/RichAsyncFunction.java]: missing empty line

File: flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java (hunk 1 above)
Comment (ifndef-SleePy) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java]: That's a better name.

File: flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java [hunk 5]
```diff
@@ -97,10 +97,16 @@ public int getNumberOfParallelSubtasks() {
 	}
 
 	@Override
+	public int getMaxNumberOfParallelSubtasks() {
+		// Number of KeyGroups is same with max-parallelism currently.
+		return taskInfo.getNumberOfKeyGroups();
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/functions/util/AbstractRuntimeUDFContext.java]: what does this return for the batch API, which doesn't have key-groups? (ideally it should ```getNumberOFParallelSubtasks()```.

**Comment by StephanEwen** [comment]
For batch jobs, I would suggest that we make sure that the `ExecutionEnvironment` or the `JobGraphGenerator` sets the max parallelism to the parallelism. Then we do not need to differentiate in the RuntimeContext.

**Comment by ifndef-SleePy** [comment]
Thank you, Stephan and zentol.
Good question and suggestion. I didn't consider it too much for batch jobs.
Also I think it's a bad idea that naming the variable "numberOfKeyGroups" in TaskInfo. Keeping max-parallelism is better. It's more common and makes much more sense for other scenarios.
I will make sure it can work with batch jobs, and update this PR later.

**Comment by StephanEwen** [comment]
I think this looks good now, merging this...

**Pull request closed without merging**